frontend/src/scenes/settings/environment/CustomChannelTypes.tsx

```diff
@@ -0,0 +1,319 @@
+            return PropertyOperator.IsNot
+        case CustomChannelOperator.IsSet:
+            return PropertyOperator.IsSet
+        case CustomChannelOperator.IsNotSet:
+            return PropertyOperator.IsNotSet
+        case CustomChannelOperator.IContains:
+            return PropertyOperator.IContains
+        case CustomChannelOperator.NotIContains:
+            return PropertyOperator.NotIContains
+        case CustomChannelOperator.Regex:
+            return PropertyOperator.Regex
+        case CustomChannelOperator.NotRegex:
+            return PropertyOperator.NotRegex
+        default:
+            throw new UnexpectedNeverError(op)
+    }
+}
+
+function keyToSessionProperty(key: CustomChannelField): string {
+    switch (key) {
+        case CustomChannelField.ReferringDomain:
+            return '$entry_referring_domain'
+        case CustomChannelField.UTMSource:
+            return '$entry_utm_source'
+        case CustomChannelField.UTMMedium:
+            return '$entry_utm_medium'
+        case CustomChannelField.UTMCampaign:
+            return '$entry_utm_campaign'
+        default:
+            throw new UnexpectedNeverError(key)
+    }
+}
+
+const sanitizeCustomChannelTypeRules = (rules: CustomChannelRule[]): CustomChannelRule[] => {
+    return (rules || [])
+        .map((rule) => {
+            return {
+                id: rule.id || uuid(),
+                channel_type: rule.channel_type,
+                combiner: rule.combiner,
+                items: (rule.items || [])
+                    .map((condition) => ({
+                        id: condition.id || uuid(),
+                        key: condition.key,
+                        op: condition.op,
+                        value: condition.value,
+                    }))
+                    .filter((item) => item.key && item.op && item.value),
+            }
+        })
+        .filter((rule) => rule.channel_type && rule.items.length > 0)
+}
+
+export function CustomChannelTypes(): JSX.Element {
+    const { updateCurrentTeam } = useActions(teamLogic)
+    const { currentTeam } = useValues(teamLogic)
+    const { reportCustomChannelTypeRulesUpdated } = useActions(eventUsageLogic)
+
+    const [savedCustomChannelTypeRules, setSavedCustomChannelTypeRules] = useState(() =>
+        sanitizeCustomChannelTypeRules(
+            currentTeam?.modifiers?.customChannelTypeRules ??
+                currentTeam?.default_modifiers?.customChannelTypeRules ??
+                []
+        )
+    )
+
+    const [customChannelTypeRules, setCustomChannelTypeRules] = useState(savedCustomChannelTypeRules)
+
+    const channelTypeOptions = useMemo((): LemonInputSelectOption[] => {
+        const optionsSet = new Set<string>()
+        customChannelTypeRules.forEach((rule) => {
+            optionsSet.add(rule.channel_type)
+        })
+        Object.values(DefaultChannelTypes).forEach((channelType) => {
+            optionsSet.add(channelType)
+        })
```

Comment:
Nitpick because I dislike `forEach`

```suggestion
        const optionsSet = new Set<string>([
           ...customChannelTypeRules.map(rule => rule.channel_type),
           ...Object.values(DefaultChannelTypes)
        ])
```